README.md
# geo-util
geo utilities for projections &amp; coordinate conversions

## Coverage

```
31 tests complete
Test duration: 45 ms
Assertions count: 90 (verbosity: 2.90)
No global variable leaks detected
Coverage: 96.13% (12/310)
lib/geographic_tiling_scheme.js missing coverage on line(s): 57, 59, 60, 124, 127, 344, 345, 348, 349, 359, 360
lib/utm_util.js missing coverage on line(s): 7
```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [v1.1.6]
- feat(GeographicTilingScheme): update `tilesForPositionsAtLevel` to optionally return only
  the tiles at the specified positions instead of all tiles in the bbox

## [v1.1.5]
- fix(utm_util): utm zones less than 10 would return incorrect EPSG code

## [v1.1.4]
- feature(GeographicTilingScheme): add `degreesToMeters` static method
- feature(GeographicTilingScheme): deprecated instance methods `toRadians` `toDegrees` and `metersToDegrees` in favor of static methods

## [v1.1.3]
- fix(GeographicTilingScheme): `tileXYToRectangle` convert string x, y, and level values to integers

## [v1.1.2]
- fix(GeographicTilingScheme): `tilesForPositionsAtLevel` was only returning the tiles on each corner. It    now returns all tiles in between as well.

## [v1.1.1]
- feat(GeographicTilingScheme): support for tmsY on all methods
- feat(GeographicTilingScheme): add `tilePixelToPosition` and `tilePixelToNativePosition` methods

## [v1.1.0]
### Added
- GeographicTilingScheme to help with tile pyramid calculations
### Changed
- Use standard.js for formatting and linting
- Use husky.js for autoformatting

## [v1.0.6]
### Added
- inverse coordinate calcuations for utm & sterea

## [v1.0.5]
### Added
- geoUtil for common geographical utility functions

## [v1.0.4]
### Added
- stereaUtil for oblique stereographic projections

## [v1.0.3]
### Changed
- fixed index.js

## [v1.0.2]
### Added
- changelog

### Changed
- fixed index.js

## [v1.0.1]
### Added
- initial commit

[1.1.0]: https://github.com/skycatch/geo-util/compare/v1.1.0...v1.0.5
[1.0.5]: https://github.com/skycatch/geo-util/compare/v1.0.5...v1.0.4
[1.0.4]: https://github.com/skycatch/geo-util/compare/v1.0.4...v1.0.3
[1.0.3]: https://github.com/skycatch/geo-util/compare/v1.0.3...v1.0.2
[1.0.2]: https://github.com/skycatch/geo-util/compare/v1.0.2...v1.0.1


test/lib/geo_util.js
'use strict'

const { expect } = require('code')
const Lab = require('lab')
const GeoUtil = require('../../lib/geo_util')
const { after, before, describe, it } = (exports.lab = Lab.script())

describe('GeoUtil', () => {
  before(() => {})
  after(() => {})

  describe('GeoUtil#getCenter', () => {
    it('returns the center', () => {
      const points = [
        { lat: 68.03753844, lon: -162.89217382 },
        { lat: 68.05426488, lon: -162.90296771 },
        { lat: 68.04542113, lon: -162.85990343 },
        { lat: 68.05604169, lon: -162.82840091 }
      ]

      const center = GeoUtil.getCenter(points)
      // lon: -162.86568431, lat: 68.04679006500001
      expect(center.lat).to.be.about(68.04679006, 0.000001)
      expect(center.lon).to.be.about(-162.865684, 0.000001)
    })
  })
})


test/lib/utm_util.js
'use strict'

const { expect } = require('code')
const Lab = require('lab')
const UtmUtil = require('../../lib/utm_util')
const { after, before, describe, it } = (exports.lab = Lab.script())

describe('UtmUtil', () => {
  before(() => {})
  after(() => {})

  describe('UtmUtil#getZone', () => {
    it('returns the UTM zone from lat/lon', () => {
      const goodPoints = [
        { lat: 37, lon: -122, zone: '10N' },
        { lat: 3, lon: 4, zone: '31N' },
        { lat: -34.5, lon: 34, zone: '36S' },
        { lat: 68.048, lon: -162.879, zone: '3N' }
      ]

      for (const point of goodPoints) {
        const zone = UtmUtil.getZone(point.lat, point.lon)
        expect(zone).to.equal(point.zone)
      }
    })

    it('throws an error on invalid lat/lon', () => {
      const invalidPoint = { lat: 'what', lon: -122 }
      const oobPoint1 = { lat: -85, lon: 4 }
      const oobPoint2 = { lat: 87, lon: 4 }

      try {
        UtmUtil.getZone(invalidPoint.lat, invalidPoint.lon)
      } catch (e) {
        expect(e.message).to.equal('Invalid point')
      }

      try {
        UtmUtil.getZone(oobPoint1.lat, oobPoint1.lon)
      } catch (e) {
        expect(e.message).to.equal('Outside UTM limits')
      }

      try {
        UtmUtil.getZone(oobPoint2.lat, oobPoint2.lon)
      } catch (e) {
        expect(e.message).to.equal('Outside UTM limits')
      }
    })
  })

  describe('UtmUtil#getProj4', () => {
    it('returns the proj4 string', () => {
      const zones = [
        {
          zone: '10N',
          proj4:
            '+proj=utm +zone=10 +ellps=WGS84 +datum=WGS84 +units=m +no_defs'
        },
        {
          zone: '10S',
          proj4:
            '+proj=utm +zone=10 +south +ellps=WGS84 +datum=WGS84 +units=m +no_defs'
        },
        {
          zone: '55n',
          proj4:
            '+proj=utm +zone=55 +ellps=WGS84 +datum=WGS84 +units=m +no_defs'
        }
      ]

      for (const zone of zones) {
        const proj4 = UtmUtil.getProj4(zone.zone)
        expect(proj4).to.equal(zone.proj4)
      }
    })

    it('throws an error on invalid utm zone', () => {
      const zones = [
        { zone: '70n' },
        { zone: '10SS' },
        { zone: 'sbn' },
        { zone: '-4s' }
      ]

      for (const zone of zones) {
        try {
          UtmUtil.getProj4(zone.zone)
          expect(true).to.equal(false)
        } catch (e) {
          expect(e.message).to.equal('UTM zone invalid')
        }
      }
    })
  })

  describe('UtmUtil#getCoord', () => {
    it('returns the UTM coordinate', () => {
      const points = [
        {
          lat: 38,
          lon: -122,
          northing: 4206286.75,
          easting: 587798.41,
          zone: '10N'
        },
        {
          lat: 32,
          lon: 94,
          northing: 3540872.53,
          easting: 594457.46,
          zone: '46N'
        },
        {
          lat: 2,
          lon: -1,
          northing: 221196.53,
          easting: 722460.63,
          zone: '30N'
        }
      ]

      for (const point of points) {
        const coord = UtmUtil.getCoord(point.lat, point.lon)
        expect(coord.northing).to.about(point.northing, 0.01)
        expect(coord.easting).to.about(point.easting, 0.01)
        expect(coord.zone).to.equal(point.zone)
      }
    })

    it('returns the UTM coordinate of given zone', () => {
      const points = [
        {
          lat: 38,
          lon: -122,
          northing: 4217628.24,
          easting: 60877.58,
          zone: '11N'
        },
        {
          lat: 32,
          lon: 94,
          northing: 3561928.3,
          easting: 1161912.96,
          zone: '45N'
        }
      ]

      for (const point of points) {
        const coord = UtmUtil.getCoord(point.lat, point.lon, point.zone)
        expect(coord.northing).to.about(point.northing, 0.01)
        expect(coord.easting).to.about(point.easting, 0.01)
        expect(coord.zone).to.equal(point.zone)
      }
    })
  })

  describe('UtmUtil#getLatLon', () => {
    it('returns the lat/lon coordinate', () => {
      const points = [
        {
          lat: 38,
          lon: -122,
          northing: 4206286.75,
          easting: 587798.41,
          zone: '10N'
        },
        {
          lat: 32,
          lon: 94,
          northing: 3540872.53,
          easting: 594457.46,
          zone: '46N'
        },
        {
          lat: 2,
          lon: -1,
          northing: 221196.53,
          easting: 722460.63,
          zone: '30N'
        }
      ]

      for (const point of points) {
        const coord = UtmUtil.getLatLon(
          point.northing,
          point.easting,
          point.zone
        )
        expect(coord.lat).to.about(point.lat, 0.000001)
        expect(coord.lon).to.about(point.lon, 0.000001)
      }
    })
  })

  describe('UtmUtil#getCustomProj4', () => {
    it('returns proj4 string for any given lat/lon', () => {
      const points = [
        {
          lat: 38,
          lon: -122,
          proj4:
            '+proj=tmerc +lat_0=38 +lon_0=-122 +k=1 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs'
        }
      ]

      for (const point of points) {
        const proj4 = UtmUtil.getCustomProj4(point.lat, point.lon)
        expect(proj4).to.equal(point.proj4)
      }
    })
  })

  describe('UtmUtil#getEPSGCode', () => {
    it('returns ESPG code for any given lat/lon', () => {
      const points = [
        { lat: 38, lon: -122, epsg: 'epsg:32610' },
        { lat: -38, lon: -122, epsg: 'epsg:32710' },
        { lat: 38, lon: -100, epsg: 'epsg:32614' },
        { lat: 68.048, lon: -162.879, epsg: 'epsg:32603' }
      ]

      for (const point of points) {
        const epsg = UtmUtil.getEPSGCode(point.lat, point.lon)
        expect(epsg).to.equal(point.epsg)
      }
    })
  })
})


test/lib/util_util.js
'use strict';

const { expect } = require('code');
const Lab = require('lab');
const UtmUtil = require('../../lib/utm_util');
const { after, before, describe, it } = exports.lab = Lab.script();

describe('UtmUtil#getZone', () => {

  before(() => {});
  after(() => {});

  it('returns the UTM zone from lat/lon', () => {

    const goodPoints = [
      { lat: 37, lon: -122, zone: '10N' },
      { lat: 3, lon: 4, zone: '31N' },
      { lat: -34.5, lon: 34, zone: '36S' }
    ];

    for (const point of goodPoints) {

      const zone = UtmUtil.getZone(point.lat, point.lon);
      expect(zone).to.equal(point.zone);
    }

  });

  it('throws an error on invalid lat/lon', () => {

    const invalidPoint = { lat: 'what', lon: -122 };
    const oobPoint1 = { lat: -85, lon: 4 };
    const oobPoint2 = { lat: 87, lon: 4 };

    try {
      UtmUtil.getZone(invalidPoint.lat, invalidPoint.lon);
    }
    catch (e) {
      expect(e.message).to.equal('Invalid point');
    }

    try {
      UtmUtil.getZone(oobPoint1.lat, oobPoint1.lon);
    }
    catch (e) {
      expect(e.message).to.equal('Outside UTM limits');
    }

    try {
      UtmUtil.getZone(oobPoint2.lat, oobPoint2.lon);
    }
    catch (e) {
      expect(e.message).to.equal('Outside UTM limits');
    }

  });

});

describe('UtmUtil#getProj4', () => {

  it('returns the proj4 string', () => {

    const zones = [
      { zone: '10N', proj4: '+proj=utm +zone=10 +ellps=WGS84 +datum=WGS84 +units=m +no_defs' },
      { zone: '10S', proj4: '+proj=utm +zone=10 +south +ellps=WGS84 +datum=WGS84 +units=m +no_defs' },
      { zone: '55n', proj4: '+proj=utm +zone=55 +ellps=WGS84 +datum=WGS84 +units=m +no_defs' }
    ];

    for (const zone of zones) {

      const proj4 = UtmUtil.getProj4(zone.zone);
      expect(proj4).to.equal(zone.proj4);
    }
  });

});


lib/sterea_util.js
'use strict';

const Proj4 = require('proj4');

module.exports = {
  /**
   * getProj4 - get proj4 string for sterea projection centered at lat/lon
   *
   * @param  {string} lat latitude of origin
   * @param  {string} lon longtitude of origin
   * @return {string}      proj4 string
   */
  getProj4: function (lat, lon) {

    return `+proj=sterea +lat_0=${lat} +lon_0=${lon} +k_0=1 +x_0=0 +y_0=0`;
  },

  /**
   * getCoord - get projected coord from lat/lon
   * using oblique stereographic projection centered at lat0/lon0
   *
   * @param  {float} lat Latitude
   * @param  {float} lon Longitude
   * @return {object}     {northing, easting, utm zone}
   */
  getCoord: function (lat, lon, lat0, lon0) {

    const proj4str = this.getProj4(lat0, lon0);

    const point = Proj4(proj4str).forward([lon, lat]);

    return {
      northing: point[1],
      easting: point[0]
    };
  },

  /**
   * getLatLon - get lat lon for given NE & origin
   *
   * @param  {float} n    northing
   * @param  {float} e    easting
   * @param  {float} lat0 latitude of origin
   * @param  {float} lon0 longitude of origin
   * @return {object}      {lat, lon}
   */
  getLatLon: function (n, e, lat0, lon0) {

    const proj4str = this.getProj4(lat0, lon0);
    const point = Proj4(proj4str).inverse([e, n]);
    return {
      lon: point[0],
      lat: point[1]
    };
  }

};


lib/geographic_tiling_scheme.js
'use strict'

const pkgjson = require('../package.json')
const debug = require('debug')(`${pkgjson.name}:geographic-tiling-scheme`)

/**
 * A tiling scheme for geometry referenced to a simple geographic projection where lat and lon
 * are directly mapped to X and Y. This projection is commonly known as geographic, equirectangular,
 * equidistant cylindrical, or plate carrée.
 *
 * This class was inpsired by the Cesium class of the same name
 * @see https://cesiumjs.org/Cesium/Build/Documentation/GeographicTilingScheme.html
 */
class GeographicTilingScheme {
  constructor (opts) {
    const defaults = {
      numOfLevelZeroTilesX: 2,
      numOfLevelZeroTilesY: 1,
      tilePixelWidth: 65, // default pixel size for terrain tiles
      tilePixelHeight: 65 // default pixel size for terrain tiles
    }
    Object.assign(this, defaults, opts)
  }

  /**
   * WGS84 ellipsoid max radius
   */
  static get ELLIPSOID_MAX_RADIUS () {
    return Math.max(6378137.0, 6378137.0, 6356752.3142451793) // WGS84 ellipsoid x,y,z in meters
  }

  /**
   * WGS84 ellipsoid circumference in meters
   */
  static get ELLIPSOID_CIRCUMFERENCE () {
    return (
      GeographicTilingScheme.ELLIPSOID_MAX_RADIUS *
      GeographicTilingScheme.TWO_PI
    )
  }

  static get TWO_PI () {
    return Math.PI * 2
  }

  static get PI_OVER_TWO () {
    return Math.PI / 2
  }

  /**
   * Largest possible rectangle for this tiling scheme
   */
  static get RECTANGLE () {
    const west = -Math.PI
    const south = -GeographicTilingScheme.PI_OVER_TWO
    const east = Math.PI
    const north = GeographicTilingScheme.PI_OVER_TWO

    const width =
      east < west ? east + GeographicTilingScheme.TWO_PI - west : east - west
    const height =
      north < south
        ? north + GeographicTilingScheme.TWO_PI - south
        : north - south

    return { west, south, east, north, width, height }
  }

  /**
   * Converts radians to degrees
   * @param {number} r Radians
   */
  static toDegrees (r) {
    return (r * 180) / Math.PI
  }

  /**
   * Converts degrees to radians
   * @param {number} d Degrees
   */
  static toRadians (d) {
    return (d * Math.PI) / 180
  }

  static metersToDegrees (m) {
    const md = GeographicTilingScheme.ELLIPSOID_CIRCUMFERENCE / 360
    return m / md
  }

  static degreesToMeters (d) {
    const md = GeographicTilingScheme.ELLIPSOID_CIRCUMFERENCE / 360
    return d * md
  }

  /**
   * @deprecated
   */
  toRadians (d) {
    debug(
      'Instance method "toRadians" is deprecated. Use static method GeographicTilingScheme.toRadians instead.'
    )
    return GeographicTilingScheme.toRadians(d)
  }

  /**
   * @deprecated
   */
  toDegrees (r) {
    debug(
      'Instance method "toDegrees" is deprecated. Use static method GeographicTilingScheme.toDegrees instead.'
    )
    return GeographicTilingScheme.toDegrees(r)
  }

  /**
   * @deprecated
   */
  metersToDegrees (m) {
    debug(
      'Instance method "metersToDegrees" is deprecated. Use static method GeographicTilingScheme.metersToDegrees instead.'
    )
    return GeographicTilingScheme.metersToDegrees(m)
  }

  /**
   * Gets the total number of tiles in the X direction at a specified level-of-detail.
   * @param {Number} level The level-of-detail.
   * @returns {Number} The number of tiles in the X direction at the given level.
   */
  getNumberOfXTilesAtLevel (level) {
    return this.numOfLevelZeroTilesX << level
  }

  /**
   * Gets the total number of tiles in the X direction at a specified level-of-detail.
   * @param {Number} level The level-of-detail.
   * @returns {Number} The number of tiles in the X direction at the given level.
   */
  getNumberOfYTilesAtLevel (level) {
    return this.numOfLevelZeroTilesY << level
  }

  /**
   * Calculates the origin coordinates of the tile containing a given lat and lon
   * @param {number} lat Latitude in decimal degrees
   * @param {number} lon Longitude in decimal degrees
   * @param {number} level The tile level-of-detail.  Zero is the least detailed.
   * @param {boolean} tmsY true to get the Y value in TMS instead of XYZ
   * @returns {object} A new object containing the tile x, y coordinates of the tile
   */
  positionToTileXY (lat, lon, level, tmsY = false) {
    const rectangle = GeographicTilingScheme.RECTANGLE
    const xTiles = this.getNumberOfXTilesAtLevel(level)
    const yTiles = this.getNumberOfYTilesAtLevel(level)
    const xTileWidth = rectangle.width / xTiles
    const yTileHeight = rectangle.height / yTiles

    lon = GeographicTilingScheme.toRadians(lon)
    lat = GeographicTilingScheme.toRadians(lat)

    let xTileCoordinate = ((lon - rectangle.west) / xTileWidth) | 0
    if (xTileCoordinate >= xTiles) xTileCoordinate = xTiles - 1

    let yTileCoordinate = ((rectangle.north - lat) / yTileHeight) | 0
    if (yTileCoordinate >= yTiles) yTileCoordinate = yTiles - 1

    if (tmsY) yTileCoordinate = yTiles - yTileCoordinate - 1

    return { x: xTileCoordinate, y: yTileCoordinate }
  }

  /**
   * Converts tile x, y coordinates and level to a cartographic rectangle in radians
   * @param {integer} x tile x origin coordinate
   * @param {integer} y tile y origin coordinate
   * @param {number} level The tile level-of-detail.  Zero is the least detailed.
   * @param {boolean} tmsY true to indicate that the Y value is TMS instead of XYZ
   */
  tileXYToRectangle (x, y, level, tmsY = false) {
    // ensure that x, y, and level are all integers
    x = ~~x
    y = ~~y
    level = ~~level

    const rectangle = GeographicTilingScheme.RECTANGLE
    const xTiles = this.getNumberOfXTilesAtLevel(level)
    const yTiles = this.getNumberOfYTilesAtLevel(level)

    const width = rectangle.width / xTiles
    const west = x * width + rectangle.west
    const east = (x + 1) * width + rectangle.west

    const height = rectangle.height / yTiles
    if (tmsY) y = yTiles - y - 1
    const north = rectangle.north - y * height
    const south = rectangle.north - (y + 1) * height

    return { west, south, east, north, width, height }
  }

  /**
   * Converts tile x, y coordinates and level to a rectangle expressed in the native coordinates of the tiling scheme
   * @param {integer} x tile x origin coordinate
   * @param {integer} y tile y origin coordinate
   * @param {number} level The tile level-of-detail.  Zero is the least detailed.
   */
  tileXYToNativeRectangle (x, y, level, tmsY = false) {
    const rect = this.tileXYToRectangle(x, y, level, tmsY)
    return this._toNativeRectangle(rect)
  }

  /**
   * Get a list of the tiles at `level` that contain the specified `positions`
   * @param {array} positions List of latlon positions
   * @param {number} level The tile level-of-detail.  Zero is the least detailed.
   * @param {boolean} tmsY true to indicate that the Y value should be in TMS instead of XYZ
   * @returns {array} Ordered list of tiles, grouped by row.
   *  The first tile in the list is the south/western most.
   *  The last tile in the list is the north/eastern most.
   */
  tilesForPositionsAtLevel (positions, level, tmsY = false) {
    let west, south, east, north

    // build the bounding box from the positions
    positions.forEach(pos => {
      const { lat, lon } = pos
      west = west === undefined || lon < west ? lon : west
      east = east === undefined || lon > east ? lon : east
      south = south === undefined || lat < south ? lat : south
      north = north === undefined || lat > north ? lat : north
    })

    const sw = { lat: south, lon: west }
    const se = { lat: south, lon: east }
    const nw = { lat: north, lon: west }
    const ne = { lat: north, lon: east }
    const bbox = [sw, se, nw, ne]

    // get the origin min/max XY values
    let minX, maxX, minY, maxY
    bbox.forEach(pos => {
      const { lat, lon } = pos
      const { x, y } = this.positionToTileXY(lat, lon, level, tmsY)

      minX = minX === undefined || minX > x ? x : minX
      maxX = maxX === undefined || maxX < x ? x : maxX
      minY = minY === undefined || minY > y ? y : minY
      maxY = maxY === undefined || maxY < y ? y : maxY
    })

    const tiles = []
    for (let offsetY = 0; offsetY <= maxY - minY; offsetY++) {
      const y = minY + offsetY
      for (let offsetX = 0; offsetX <= maxX - minX; offsetX++) {
        const x = minX + offsetX
        tiles.push({ x, y })
      }
    }

    return tiles
  }

  /**
   * Compute the rectangle (bounding box) in radians that covers the tiles needed to contain the specified
   * positions at the specified tile level-of-detail
   * @param {array} positions List of { lat, lon } positions
   * @param {number} level Tile level-of-detail
   * @param {number} offset Uniform offset (in meters) to apply to the bounds, use negative to max the rectangle smaller
   * @param {boolean} tmsY true to indicate that the Y value should be in TMS instead of XYZ
   * @returns {object} Rectangle in radians
   */
  rectangleForPositionsAtLevel (positions, level, offset = 0, tmsY = false) {
    const tiles = this.tilesForPositionsAtLevel(positions, level, tmsY)
    // the first and last tiles will contain the min/max XY origins
    // we can safely ignore all the tiles in between
    const firstLast = [tiles[0], tiles[tiles.length - 1]]

    // calculate the encompassing rectangle
    const res = {}

    firstLast.forEach(tile => {
      const rect = this.tileXYToRectangle(tile.x, tile.y, level, tmsY)
      const { west, south, east, north } = rect
      res.west = res.west === undefined || west < res.west ? west : res.west
      res.east = res.east === undefined || east > res.east ? east : res.east
      res.south =
        res.south === undefined || south < res.south ? south : res.south
      res.north =
        res.north === undefined || north > res.north ? north : res.north
    })

    // Bring in all side by the maximum allowed geometric error to avoid accidentally
    // bleeding over the tile boundary. Cesium allows for some gap between tiles
    if (offset) {
      offset = GeographicTilingScheme.metersToDegrees(offset)
      res.west = res.west - offset
      res.east = res.east + offset
      res.south = res.south - offset
      res.north = res.north + offset
    }

    const width = res.east - res.west
    const height = res.north - res.south

    return { ...res, width, height }
  }

  /**
   * Compute the rectangle (bounding box) in degrees that covers the tiles needed to contain the specified
   * positions at the specified tile level-of-detail
   * @param {array} positions List of { lat, lon } positions
   * @param {number} level Tile level-of-detail
   * @param {number} offset Uniform offset (in meters) to apply to the bounds, use negative to max the rectangle smaller
   * @param {boolean} tmsY true to indicate that the Y value should be in TMS instead of XYZ
   * @returns {object} Rectangle in degrees
   */
  nativeRectangleForPositionsAtLevel (
    positions,
    level,
    offset = 0,
    tmsY = false
  ) {
    const rect = this.rectangleForPositionsAtLevel(
      positions,
      level,
      offset,
      tmsY
    )
    return this._toNativeRectangle(rect)
  }

  /**
   * Convert a rectangle in radians to degrees
   * @param {object} rectangle Rectangle in radians
   * @returns {object} Rectangle in degrees
   */
  _toNativeRectangle ({ west, south, east, north, width, height }) {
    west = GeographicTilingScheme.toDegrees(west)
    south = GeographicTilingScheme.toDegrees(south)
    east = GeographicTilingScheme.toDegrees(east)
    north = GeographicTilingScheme.toDegrees(north)
    width = GeographicTilingScheme.toDegrees(width)
    height = GeographicTilingScheme.toDegrees(height)
    return { west, south, east, north, width, height }
  }

  /**
   * Gets the maximum geometric error allowed in a tile at a given level.
   * {@link_https://github.com/AnalyticalGraphicsInc/cesium/blob/1.58/Source/Core/TerrainProvider.js}
   * @param {number} level The tile level for which to get the maximum geometric error.
   * @returns {number} The maximum geometric error in meters
   */
  getLevelMaximumGeometricError (level) {
    const tileImageWidth = this.tilePixelWidth
    const heightmapTerrainQuality = 0.25
    const numberOfTilesAtLevelZero = this.numOfLevelZeroTilesX
    const ellpsoidMaximumRadius = GeographicTilingScheme.ELLIPSOID_MAX_RADIUS
    const levelZeroMaximumGeometricError =
      (ellpsoidMaximumRadius *
        GeographicTilingScheme.TWO_PI *
        heightmapTerrainQuality) /
      (tileImageWidth * numberOfTilesAtLevelZero)

    return levelZeroMaximumGeometricError / (1 << level)
  }

  /**
   * Get the geographic x,y in radians for the specified pixel coordinate.
   * Pixel coordinate origin (0,0) is expected to be the bottom left (southwest) corner.
   * The `flipY` flag can be used to set the pixel origin to the top left (northwest) corner.
   * @param {object} rectangle Rectangle in radians
   * @param {number} pixelX
   * @param {number} pixelY
   * @param {boolean} flipY True to flip the Y axis
   */
  tilePixelToPosition (rectangle, pixelX, pixelY, flipY = false) {
    const { west, south, north, width, height } = rectangle

    if (pixelX < 0 || pixelX > this.tilePixelWidth - 1) {
      throw new Error('PixelX is out of bounds')
    }

    if (pixelY < 0 || pixelY > this.tilePixelHeight - 1) {
      throw new Error('PixelY is out of bounds')
    }

    // assumes pixel origin is bottom left (sw)
    // assumes pixel position is zero indexed
    const xPercent = pixelX / (this.tilePixelWidth - 1)
    const yPercent = pixelY / (this.tilePixelHeight - 1)
    const xPos = west + xPercent * width
    let yPos = south + yPercent * height

    if (flipY) {
      yPos = north - yPos
    }

    return { x: xPos, y: yPos }
  }

  /**
   * Get the geographic x,y in latlon for the specified pixel coordinate
   * @param {object} rectangle Rectangle in radians
   * @param {number} pixelX
   * @param {number} pixelY
   * @param {boolean} flipY True to flip the Y axis
   */
  tilePixelToNativePosition (rectangle, pixelX, pixelY, flipY = false) {
    let { x, y } = this.tilePixelToPosition(rectangle, pixelX, pixelY, flipY)
    x = GeographicTilingScheme.toDegrees(x)
    y = GeographicTilingScheme.toDegrees(y)
    return { x, y }
  }
}

module.exports = GeographicTilingScheme


lib/utm_util.js
'use strict'

const Proj4 = require('proj4')

module.exports = {
  getZone: function (lat, lon) {
    if (isNaN(lat) || isNaN(lon)) {
      throw new Error('Invalid point')
    }

    // typecast
    lat = parseFloat(lat)
    lon = parseFloat(lon)

    if (!(lat > -80 && lat <= 84)) {
      throw new Error('Outside UTM limits')
    }

    const zone = Math.floor((lon + 180) / 6) + 1 // longitudinal zone

    if (lat < 0) {
      return `${zone}S`
    }

    return `${zone}N`
  },

  /**
   * getProj4 - get proj4 string for UTM zone
   *
   * @param  {string} zone utm zone name
   * @return {string}      proj4 string
   */
  getProj4: function (zone) {
    if (zone.length > 3) {
      throw new Error('UTM zone invalid')
    }

    const isSouth = zone.substr(-1, 1).toLowerCase() === 's'
    zone = parseInt(zone)

    if (isNaN(zone) || zone < 1 || zone > 60) {
      throw new Error('UTM zone invalid')
    }

    if (isSouth) {
      return `+proj=utm +zone=${zone} +south +ellps=WGS84 +datum=WGS84 +units=m +no_defs`
    }

    return `+proj=utm +zone=${zone} +ellps=WGS84 +datum=WGS84 +units=m +no_defs`
  },

  /**
   * getCoord - get UTM coord from lat/lon
   *
   * @param  {float} lat Latitude
   * @param  {float} lon Longitude
   * @return {object}     {northing, easting, utm zone}
   */
  getCoord: function (lat, lon, zone = undefined) {
    // calc zone if not provided
    if (!zone) {
      zone = this.getZone(lat, lon)
    }
    const proj4str = this.getProj4(zone)

    const point = Proj4(proj4str).forward([lon, lat])

    return {
      northing: point[1],
      easting: point[0],
      zone
    }
  },

  /**
   * getLatLon - get lat/lon from UTM coordinate
   *
   * @param  {type} n    northing
   * @param  {type} e    easting
   * @param  {type} zone UTM zone
   * @return {type}      {lat, lon}
   */
  getLatLon: function (n, e, zone) {
    const proj4str = this.getProj4(zone)

    const point = Proj4(proj4str).inverse([e, n])

    return {
      lon: point[0],
      lat: point[1]
    }
  },

  /**
   * getCustomProj4 - get custom proj4 string
   *
   * @param  {float} lat Latitude
   * @param  {float} lon Longitude
   * @return {string}     proj4 string
   */
  getCustomProj4: function (lat, lon) {
    return `+proj=tmerc +lat_0=${lat} +lon_0=${lon} +k=1 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs`
  },

  /**
   * getEPSGCode - Returns the WGS84 UTM Zone EPSG code for a give latlon
   *
   * @param  {float} lat Latitude
   * @param  {float} lon Longitude
   * @return {string}     EPSG string
   * @see http://gis.stackexchange.com/questions/33219/how-to-convert-lat-long-to-utm-using-proj4j-similar-to-jscience-utm-latlongtou
   */
  getEPSGCode: function (lat, lon) {
    const zone = this.getZone(lat, lon)
    const zoneNum = zone.replace('N', '').replace('S', '')

    if (zone.indexOf('S') !== -1) {
      return `epsg:327${zoneNum}`
    }

    return `epsg:326${zoneNum}`
  }
}
